autoload/lsp/internal/inlay_hints.vim

```diff
@@ -0,0 +1,127 @@
+let s:use_vim_textprops = lsp#utils#_has_vim_virtual_text() && !has('nvim')
+
+function! s:set_inlay_hints(data) abort
+    let l:bufnr = bufnr('%')
+
+    call s:clear_inlay_hints()
+
+    if mode() !=# 'n' | return | endif
+
+    if lsp#client#is_error(a:data['response']) | return | endif
+
+    " Get hints from the response
+    let l:hints = a:data['response']['result']
+    if empty(l:hints)
+        return
+    endif
+
+    let l:not_curline = s:has_inlay_hints_mode('!curline')
+    for l:hint in l:hints
+        if l:not_curline && l:hint.position.line+1 ==# line('.')
+            continue
+        endif
+        let l:label = ''
+        if type(l:hint.label) ==# v:t_list
+            let l:label = join(map(copy(l:hint.label), {_,v -> v.value}), ', ')
+        else
+            let l:label = l:hint.label
+        endif
+        let l:text = (get(l:hint, 'paddingLeft', v:false) ? ' ' : '') . l:label . (get(l:hint, 'paddingRight', v:false) ? ' ' : '')
+        if !has_key(l:hint, 'kind') || l:hint.kind ==# 1
+            call prop_add(l:hint.position.line+1, l:hint.position.character+1, {'type': 'vim_lsp_inlay_hint_type', 'text': l:text, 'bufnr': l:bufnr})
+        elseif l:hint.kind ==# 2
+            call prop_add(l:hint.position.line+1, l:hint.position.character+1, {'type': 'vim_lsp_inlay_hint_parameter', 'text': l:text, 'bufnr': l:bufnr})
```

Comment:
can use `prop_add_list` instead for perf. We did this for semantic highlight.

Reply:
I think it is difficult to use prop_add_list for this since inlay_hints include text attribute for each props instead of semantic highlight.

Reply:
we should be good for now.